teps/0031-tekton-bundles-cli.md

```diff
@@ -66,3 +71,95 @@ Provide a way for users to...
 * Users may publish to any repository they have access to.
 
 * Only valid Tekton Bundles are produced.
+
+## Proposal
+
+Amend the Tekton CLI to provide the following commands (more detail later):
+
+```shell
+# REF = Docker-like Tekton Bundle reference, eg. docker.io/myworkspace/mybundle:1.0
+# BUNDLE_OBJECT = path to a file or directory or STDIN of a valid Tekton Task or Pipeline object to be included in the
+# bundle.
+
+# Generates a Tekton Bundle from a set of 1+ bundle objects. Like `docker`, tags the object with the specified reference
+# locally.
+tkn remote build <REF> [BUNDLE_OBJECT...]
+
+# Like docker, publishing the locally referenced Tekton Bundle to a remote repository.
+tkn remote push <REF>
+
+# Fetches a Tekton Bundle and prints its contents in a configurable format. If KIND is specified, will print only
+# objects of the specified kind. If KIND and NAME are specified, will retrieve a specific object.
+tkn remote get <REF> [KIND] [NAME] --output=[FORMAT]
+
+# The following are amendments to existing commands.
+
+# Retrieves from an image the requested task(s) rather than from the cluster.
+tkn task list --image=<REF>
+tkn task get --image=<REF> [NAME]
+
+# Adds the specified BUNDLE_OBJECT to the referenced Tekton Bundle.
+tkn task create --image=<REF> [BUNDLE_OBJECT]
```

Comment:
If we have this — being able to add a task to a tekton bundle — we may not need `tkn remote`.

Reply:
+1 and same goes for `tktn bundle push` would it be possible to fit this feature in a `tkn task push` instead ?

Reply:
This is an interesting question. My reasoning for having a `bundle` or `remote` command was with the idea that the "average" Tekton Bundle would be a `Pipeline` and the `Task`s in that pipeline in one bundle. This way, all of the dependencies of a single pipeline run are versioned. With that use case in mind, I figured it would make sense to have a command to publish an entire bundle of mixed objects at once.

The reason for using the `tkn task` and `tkn pipeline` commands as well (in addition to the build + push) was so that this bundle could be modified before/after publishing.

If we remove the `bundle` command, how would we allow users to create a "complete" bundle?

Reply:
@chmouel @vdemeester I find using `task` here is confusing instead.  `[BUNDLE_OBJECT]` can be more than a single task?  This looks like a `bundle` command to me.

I'm not sure about `create` either:

- If this a way to bypass the build (tag) + push workflow, I would not do it. If we provide two ways to do the same thing it will become confusing. 
- If this is about adding content to an existing bundle, the command should probably be called `add`... however, adding content sounds like awfully complex, because we'd have to deal with conflicts

Reply:
If the aim of this is to add a single task to an existing bundle, we could have:
```
tkn task add-to-bundle --bundle=<REF> [FILE]
```

which could provide a quick alternative for adding a tasks to bundles without building a folder with README etc. We could not guarantee that `[FILE]` includes a single task, but I guess that would be acceptable.

If we go down this route though, we would also fail if `[FILE]` includes pipelines, and we should also have `tkn pipeline` commands to work with bundles. I'm not sure we should do that really?